lib/soak/src/bin/observer.rs

```diff
@@ -97,123 +131,233 @@ enum Status {
+        let mut wtr = BufWriter::new(fs::File::create(self.capture_path).await?);
+        while let Some(msg) = self.rcv.recv().await {
+            wtr.write_all(msg.as_bytes()).await?;
+            wtr.write_all(b"\n").await?;
+        }
+        info!("All sender channels closed, flushing writer and exiting.");
+        wtr.flush().await?;
+        Ok(())
+    }
 }
 
-struct Worker {
-    vector_id: String,
+struct TargetWorker {
+    snd: Sender<String>,
     experiment_name: String,
+    target_id: String,
+    url: Url,
+    run_id: Arc<Uuid>,
     variant: soak::Variant,
-    queries: Vec<(Query, Url)>,
-    capture_path: String,
 }
 
-impl Worker {
-    fn new(config: Config) -> Self {
-        let mut queries = Vec::with_capacity(config.queries.len());
-        for query in config.queries {
-            let url = format!("{}/api/v1/query?query={}", config.prometheus, query.query);
-            let url = url.parse::<Url>().unwrap();
-            queries.push((query, url));
-        }
+impl fmt::Debug for TargetWorker {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("TargetWorker")
+            .field("run_id", &self.run_id)
+            .field("target_id", &self.target_id)
+            .field("url", &self.url.as_str())
+            .finish()
+    }
+}
 
+impl TargetWorker {
+    fn new(
+        snd: Sender<String>,
+        experiment_name: String,
+        target_id: String,
+        url: Url,
+        run_id: Arc<Uuid>,
+        variant: soak::Variant,
+    ) -> Self {
         Self {
-            vector_id: config.vector_id,
-            experiment_name: config.experiment_name,
-            variant: config.variant,
-            queries,
-            capture_path: config.capture_path,
+            snd,
+            experiment_name,
+            target_id,
+            url,
+            run_id,
+            variant,
         }
     }
 
+    #[instrument]
     async fn run(self) -> Result<(), Error> {
         let client: reqwest::Client = reqwest::Client::new();
 
-        let file = fs::File::create(self.capture_path)?;
-        let mut wtr = csv::Writer::from_writer(file);
         for fetch_index in 0..u64::max_value() {
-            for (query, url) in &self.queries {
-                let request = client.get(url.clone()).build()?;
-                let body = client
-                    .execute(request)
-                    .await?
-                    .json::<QueryResponse>()
-                    .await?;
-                debug!("body: {:?}", body.data);
-
-                if !body.data.result.is_empty() {
-                    let time = body.data.result[0].value.time;
-                    let value = body.data.result[0].value.value.parse::<f64>()?;
+            if TERMINATE.load(Ordering::Relaxed) {
+                info!("Received terminate signal");
+                break;
+            }
+            sleep(Duration::from_millis(307)).await;
```

Comment:
It's a smallish prime. This timing loop doesn't introspect on how delayed it is, so I wanted something that wouldn't tend to be a multiple of anything else, was less than 1000 and what not. Completely arbitrary otherwise.